designs/2025-rule-languages/README.md

```diff
@@ -0,0 +1,119 @@
+- Repo: eslint/eslint
+- Start Date: 2025-06-11
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Nicholas C. Zakas
+
+# Allow rules to specify the languages/dialects they work on
+
+## Summary
+
+This RFC proposes adding metadata to ESLint rules that indicate which programming languages and language dialects they support. This will enable better documentation and runtime capabilities for ESLint when working with multiple languages.
+
+## Motivation
+
+Currently, an ESLint rule has no way to indicate which languages or language dialects it is designed to work with. This limitation creates two significant problems:
+
+1. **Documentation purposes** - Users have no easy way to determine which JavaScript rules have been updated to support TypeScript syntax, for example. This makes it harder to understand which rules can be safely enabled when linting TypeScript code.
+
+2. **Runtime purposes** - ESLint cannot automatically disable rules that don't apply to a given language. For instance, when linting CSS files, JavaScript-specific rules should ideally be automatically turned off, but currently there's no mechanism to do this.
+
+As ESLint's ecosystem expands to support more languages beyond JavaScript (such as TypeScript, CSS, and potentially others), having a standardized way to specify language compatibility becomes increasingly important for both users and maintainers.
+
+## Detailed Design
+
+I propose adding a `languages` array to the rule's `meta` object. Example:
+
+```js
+// Rule definition example
+module.exports = {
+  meta: {
+    type: "suggestion",
+    docs: {
+      description: "Disallow something",
+      recommended: true
+    },
+    
+    // new property
+    languages: ["markdown/gfm", "markdown/commonmark"] // Languages the rule supports
+  },
+  create(context) {
+    // Rule implementation
+    return {};
+  }
+};
+```
+
+The `languages` array will contain strings that identify the specific language plugins the rule is designed to work with. Each string follows the standardized format `"plugin/language"` to uniquely identify the language plugin. To specify that a rule works with any language in a plugin, the format of `"plugin/*"` is used.
```

Comment:
We can also support `"*"` to specify that a rule works with any language (e.g., `no-restricted-syntax`).